### PR TITLE
UEFI: complete rewrite to support EFI boot

### PR DESCRIPTION
**Pull request by pirat89**
The original solution broke the upgrade and it's one of reasons why
    the upgrade has been inhibited for EFI boot. This should handle the
    EFI boot during the upgrade, however manual actions after the upgrade
    will be required regarding the Legacy Grub -> Grub2 migration.
    
    After discussion with the bootlaoder team,
    - drop the stuff around 'shim' rpm as secure boot is not supported
      on RHEL 6
    - EFI boot is relevant just on the intel architecture in case
      of RHEL 6
    
    Short summary of actions
    - check whether EFI boot is used
    - backup & restore Legacy Grub EFI files
    - ensure the next boot is done to the correct partition

### CODE REVIEW

**Comment by pirat89** [comment]
Note: the documentation is not updated yet, to inform about manual steps after the upgrade related to EFI. It will be prepared later in the mentioned document (see the solution text)

**Comment by pirat89** [comment]
@jhlavac ping :)

File: RHEL6_7/system/uefi/check [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+
+is_uefi || exit_not_applicable
+
+# I do not remember the exact reason, but we want to rather check the
+# executable exists instead of the rpm in this case
+command -v efibootmgr >/dev/null || {
+  log_extreme_risk "The EFI has been detected but efibootmgr is missing"
+  msg="The EFI firmware has been detected but the efibootmgr utility."
+  msg+=" is missing. Without the utility it is not possible to proceed"
+  msg+=" the in-place upgrade with EFI. Install the efibootmgr rpm and run"
```
Comment (bocekm) [inline RHEL6_7/system/uefi/check]:
```suggestion
  msg="The EFI firmware has been detected but the efibootmgr utility"
  msg+=" is missing. Without the utility it is not possible to proceed with"
  msg+=" the in-place upgrade. Install the efibootmgr rpm and run"
```

File: RHEL6_7/system/uefi/check [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
+
+# create pre-upgrade and post-upgrade scripts
+cp -a "efibootorder_fix.sh" "$PREUPGRADE_SCRIPT_DIR/"
+cp -a "efibootorder_fix.sh" "$POSTUPGRADE_DIR/"
+chmod +x "$PREUPGRADE_SCRIPT_DIR/efibootorder_fix.sh"
+chmod +x "$POSTUPGRADE_DIR/efibootorder_fix.sh"
 
-echo "This system does not use the EFI boot loader. The Preupgrade Assistant will not replace your current boot loader automatically as it is too dangerous. If you want to use GRUB 2, do it manually after the upgrade by using 'grub2-install' and 'grub2-mkconfig'." > solution.txt
-exit_informational
+log_medium_risk "The Legacy GRUB has to be migrated to GRUB 2 after the upgrad emanually"
```
Comment (bocekm) [inline RHEL6_7/system/uefi/check]:
```suggestion
log_medium_risk "The Legacy GRUB has to be migrated to GRUB 2 after the upgrade manually"
```

File: RHEL6_7/system/uefi/check [hunk 3]
```diff
@@ -1,14 +1,33 @@
...
+exit_fail
```
Comment (bocekm) [inline RHEL6_7/system/uefi/check:33]: I hope the table in https://github.com/upgrades-migrations/preupgrade-assistant/wiki/How-modules-affect-the-Preupgrade-Assistant-return-code still works:
```
| exit_x         | log_x_risk    | result in XML    | return code |
|================|===============|==================|=============|
| fail           | medium        | needs_inspection |  0          |
```

File: RHEL6_7/system/uefi/module.ini [hunk 4]
```diff
@@ -1,5 +1,4 @@
 [preupgrade]
-content_description = The module checks for EFI and it forbids an in-place upgrade when the EFI boot loader is present.
-content_title = UEFI boot loader
-author = Frantisek Kluknavsky <fkluknav@redhat.com>
-binary_req: efibootmgr
+author = Petr Stodulka <pstodulk@redhat.com>
+content_title = Handle UEFI during the in-place upgrade
+content_description = Set up the next boot entry correctly on EFI systems to ensure the boot in correct partition will happen after each restart during the in-place upgrade procedure.
```
Comment (bocekm) [inline RHEL6_7/system/uefi/module.ini]:
```suggestion
content_description = Set up the next boot entry correctly on EFI systems to ensure that booting will happen in a correct partition after each restart during the in-place upgrade procedure.
```

File: RHEL6_7/system/uefi/efibootorder_fix.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+log_info "Check the EFI:"
+efibootmgr || {
+    log_error "Something is wrong. Cannot use the efibootmgr utility"
+    exit 1
+}
+
+# the $efibin_path is removed during the upgrade as it is provided by the grub
+# rpm, that is removed during the upgrade
```
Comment (bocekm) [inline RHEL6_7/system/uefi/efibootorder_fix.sh]:
```suggestion
# rpm which is removed during the upgrade
```

File: RHEL6_7/system/uefi/efibootorder_fix.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+efibin_path="/boot/efi/EFI/redhat/grub.efi"
+eficonf_path="/boot/efi/EFI/redhat/grub.conf"
+if [ -e "$efibin_path" ] ; then
+    # now we are for sure on the original system, as the binary is not removed
+    # backup these files
+    log_info "Back up EFI files."
```
Comment (bocekm) [inline RHEL6_7/system/uefi/efibootorder_fix.sh]:
```suggestion
    log_info "Backing up EFI files."
```

File: RHEL6_7/system/uefi/efibootorder_fix.sh [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    cp -a ${efibin_path}{,.preupg}
+    cp -a ${eficonf_path}{,.preupg}
+else
+    # restore the files from the backup
+    log_info "Restore EFI files."
```
Comment (bocekm) [inline RHEL6_7/system/uefi/efibootorder_fix.sh]:
```suggestion
    log_info "Restoring EFI files."
```

File: RHEL6_7/system/uefi/efibootorder_fix.sh [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    log_error "Cannot detect the current EFI boot using the efibootmgr"
+    exit 1
+}
+log_info "The current EFI boot is: $current_boot"
+
+# e.g. BootNext: 0001
+next_boot=$(efibootmgr | grep "^BootNext:" | cut -d ":" -f 2- | sed -r "s/^\s*(.*)\s*$/\1/" | grep -o "^[0-9A-F]*")
+[ -z "$next_boot" ] && {
+    # We set BootNext to CurrentBoot only if BootNext wasn't previously set
+    log_info "Set the next boot to: $current_boot"
```
Comment (bocekm) [inline RHEL6_7/system/uefi/efibootorder_fix.sh]:
```suggestion
    log_info "Setting the next boot to: $current_boot"
```

**Review by bocekm — CHANGES_REQUESTED** [review]

File: RHEL6_7/system/uefi/check (hunk 3 above)
Comment (pirat89) [reply RHEL6_7/system/uefi/check:33]: yes :)

**Review by bocekm — APPROVED** [review]

**Pull request merged**